framework/stats/stats.cpp

```diff
@@ -279,6 +288,149 @@ void Stats::push_sample(const StatsProvider::Counters &sample)
+		case StatIndex::gpu_cycles:
+			return "GPU Cycles (M/s)";
+		case StatIndex::gpu_vertex_cycles:
+			return "Vertex Cycles (M/s)";
+		case StatIndex::gpu_load_store_cycles:
+			return "Load Store Cycles (k/s)";
+		case StatIndex::gpu_tiles:
+			return "Tiles (k/s)";
+		case StatIndex::gpu_killed_tiles:
+			return "Tiles killed by CRC match (k/s)";
+		case StatIndex::gpu_fragment_jobs:
+			return "Fragment Jobs (s)";
+		case StatIndex::gpu_fragment_cycles:
+			return "Fragment Cycles (M/s)";
+		case StatIndex::gpu_tex_cycles:
+			return "Shader Texture Cycles (k/s)";
+		case StatIndex::gpu_ext_reads:
+			return "External Reads (M/s)";
+		case StatIndex::gpu_ext_writes:
+			return "External Writes (M/s)";
+		case StatIndex::gpu_ext_read_stalls:
+			return "External Read Stalls (M/s)";
+		case StatIndex::gpu_ext_write_stalls:
+			return "External Write Stalls (M/s)";
+		case StatIndex::gpu_ext_read_bytes:
+			return "External Read Bytes (MiB/s)";
+		case StatIndex::gpu_ext_write_bytes:
+			return "External Write Bytes (MiB/s)";
+		default:
+			return nullptr;
+	}
+}
+}        // namespace
+
+void Stats::profile_counters() const
+{
+	static std::chrono::high_resolution_clock::time_point last_time = std::chrono::high_resolution_clock::now();
+	std::chrono::high_resolution_clock::time_point        now       = std::chrono::high_resolution_clock::now();
+
+	if (now - last_time < std::chrono::milliseconds(100))
+	{
+		return;
+	}
+
+	last_time = now;
+
+	for (auto &c : counters)
+	{
+		StatIndex idx        = c.first;
+		auto     &graph_data = get_graph_data(idx);
+
+		if (c.second.empty())
+		{
+			continue;
+		}
+
+		float average = std::accumulate(c.second.begin(), c.second.end(), 0.0f) / c.second.size();
+
+		if (auto *index_name = to_string(idx))
+		{
+			Plot<float>::plot(index_name, average * graph_data.scale_factor);
+		}
+	}
+
+	static std::vector<std::string> labels;
+
+	auto        &device    = render_context.get_device();
+	VmaAllocator allocator = device.get_memory_allocator();
+
+	VmaBudget heap_budgets[VK_MAX_MEMORY_HEAPS];
+	vmaGetHeapBudgets(allocator, heap_budgets);
+
+	// We know that we will only ever have one device in the system, so we can cache the labels
+	if (labels.size() == 0)
+	{
+		VkPhysicalDeviceMemoryProperties memory_properties;
+		vkGetPhysicalDeviceMemoryProperties(device.get_gpu().get_handle(), &memory_properties);
+
+		labels.reserve(memory_properties.memoryHeapCount);
+
+		for (size_t heap = 0; heap < memory_properties.memoryHeapCount; heap++)
+		{
+			VkMemoryPropertyFlags flags = 0;
+			for (size_t type = 0; type < memory_properties.memoryTypeCount; type++)
+			{
+				if (memory_properties.memoryTypes[type].heapIndex == heap)
+				{
+					flags |= memory_properties.memoryTypes[type].propertyFlags;
+				}
+			}
+
+			// TODO: Using VulkanHPP to_string here instead of vkb::to_string. After deciding on the framework direction we should adjust this
+			std::stringstream ss;
```

Comment:
You don't have to use a `std::stringstream`:
`labels.push_back("Heap " + std::to_string(heap) + " " + vk::to_string(vk::MemoryPropertyFlags{flags}));
`